### PR TITLE
express-notes-api | Samuel Cho

### PR DESCRIPTION
**Pull request by Samuel-Cho**
Built a small API to create, read, update, and delete notes.
Screen Shot 2021-03-16 at 5 30 22 PM
Screen Shot 2021-03-16 at 7 29 34 PM
Screen Shot 2021-03-16 at 7 30 14 PM
Screen Shot 2021-03-16 at 7 30 43 PM
Screen Shot 2021-03-16 at 7 30 56 PM
Screen Shot 2021-03-16 at 8 02 45 PM
Screen Shot 2021-03-16 at 8 03 00 PM
Screen Shot 2021-03-16 at 8 03 13 PM
Screen Shot 2021-03-16 at 8 10 28 PM
Screen Shot 2021-03-16 at 8 25 36 PM
Screen Shot 2021-03-16 at 8 26 13 PM
Screen Shot 2021-03-16 at 8 32 17 PM
Screen Shot 2021-03-16 at 8 34 12 PM
Screen Shot 2021-03-16 at 9 04 20 PM
Screen Shot 2021-03-16 at 9 04 59 PM
Screen Shot 2021-03-16 at 9 05 48 PM
Screen Shot 2021-03-16 at 9 07 14 PM
Screen Shot 2021-03-16 at 9 08 07 PM

### CODE REVIEW

File: express-notes-api/deleteNote.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  if (Math.sign(req.params.id) !== 1) {
+    res.status(400).json(error1);
+  } else {
+    if (data.notes.hasOwnProperty(req.params.id)) {
+      delete data.notes[req.params.id];
+      fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf8', (err) => {
+        if (err) {
+          res.status(500).json(error2);
+        }
+        res.status(204).end();
```
Comment (thebearingedge) [inline express-notes-api/deleteNote.js:16]: Be sure that it is impossible for the server to try to respond more than once to the same request. You'll need an `else` in here.

File: express-notes-api/postNote.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    res.status(400).json(error1);
+  } else {
+    req.body.id = data.nextId;
+    data.notes[data.nextId] = req.body;
+    data.nextId++
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf8', (err) => {
+      if (err) {
+        res.status(500).json(error2);
+      }
+      res.status(201).json(req.body);
```
Comment (thebearingedge) [inline express-notes-api/postNote.js:17]: Be sure that it is impossible for the server to try to respond more than once to the same request. You'll need an `else` in here.

File: express-notes-api/putNote.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+      "error": `cannot find note with id ${req.params.id}`
+    };
+    res.status(404).json(error4);
+  } else {
+    data.notes[req.params.id].content = req.body.content;
+    fs.writeFile('data.json', JSON.stringify(data, null, 2), 'utf8', (err) => {
+      if (err) {
+        res.status(500).json(error3);
+      }
+      res.status(200).json(data.notes[req.params.id]);
```
Comment (thebearingedge) [inline express-notes-api/putNote.js:23]: Be sure that it is impossible for the server to try to respond more than once to the same request. You'll need an `else` in here.

**Review by thebearingedge — CHANGES_REQUESTED** [review]

**Comment by Samuel-Cho** [comment]
Screen Shot 2021-03-17 at 3 25 40 PM
Screen Shot 2021-03-17 at 3 26 17 PM
Screen Shot 2021-03-17 at 3 26 53 PM

**Review by thebearingedge — APPROVED** [review]

**Pull request merged**